From 55371cca1bb4d513168c5802274f2b0273ee53d7 Mon Sep 17 00:00:00 2001
From: Emmanuel Jillela <emmanuel.jillela@intel.com>
Date: Tue, 9 Feb 2016 14:03:39 -0700
Subject: [PATCH 258/441]      Title: Invoke the ACPI control method to
 indicate BBU service level      Issue: Jira: CPM-247      IssueType: Feature
     Resolution: Notify acpi whether internal BBU is used or not.      Impact:
      Change-Id: I53c26af5b72ff576664de55616c6f41c05db9ec4

---
 arch/x86/NetIP_SubSystem/netip_subsystem_pm.c | 38 +++++++++++++++++++++++++++
 1 file changed, 38 insertions(+)

--- a/arch/x86/NetIP_SubSystem/netip_subsystem_pm.c
+++ b/arch/x86/NetIP_SubSystem/netip_subsystem_pm.c
@@ -119,6 +119,9 @@ static bool g_netss_boot_handshake_done
 static struct timer_list netss_pm_timer;
 static bool g_timer_created = false;
 
+/**Need to notify once whether internal BBU is on or off */
+static bool g_acpi_notified_bbu_status = false;
+
 /**
  * puma_netip_notify_state() - invoke to update the state of NETIP
  * @state: state of NETIP, 0 - Active, 1 - BBU, 2 - Standby, 3 - Deep standby
@@ -166,6 +169,30 @@ static void netss_handle_resource_power_
    }
    return;
 }
+static void netss_indicate_bbu_service_level(uint8_t level)
+{
+   struct acpi_object_list input;
+   union acpi_object in_param;
+   input.count = 1;
+   input.pointer = &in_param;
+   in_param.type = ACPI_TYPE_INTEGER;
+   in_param.integer.value = level;
+
+   if(pnetss_drv_data->acpi_h == NULL)
+   {
+      printk("NetIP acpi handle is NULL file: %s line: %d\n", __FILE__, __LINE__);
+   }
+   else
+   {
+      /** call ACPI function from here to inform whether internal BBU is being used */
+      if(!g_acpi_notified_bbu_status)
+      {
+         NETSS_PM_PRINT("Internal BBU being used? %s \n", (level? "YES" : "NO"));
+         acpi_evaluate_object(pnetss_drv_data->acpi_h, "NBED", &input, NULL);
+         g_acpi_notified_bbu_status= true;
+      }
+   }
+}
 static void netss_pm_timeout_handler (unsigned long data)
 {
    printk("IT APPEARS THAT ATOM NOT RECEIVING MSGS FROM NETIP \n");
@@ -283,6 +310,12 @@ static void ness_process_netip_props_msg
       /*Check next two bytes */
       data++;
    }
+   /**Internal BBU service does not exist */
+   if(!(g_netip_services_available & (1 << (NETSS_SERVICE_ID_BBU-1))))
+   {
+      /**Inform ACPI internal BBU service is off */
+      netss_indicate_bbu_service_level(0);
+   }
    NETSS_PM_PRINT("services that are available = %x\n", g_netip_services_available);
 
    /*Remove the sysfs attribute files of the services that are not available */
@@ -441,6 +474,11 @@ static void netss_update_sysfs_attribute
       {
          netss_handle_resource_power_state(service_msg->service[i].id, service_msg->service[i].state);
       }
+      /**Internal BBU service is available, inform acip whether user turning it on or off */
+      if(service_msg->service[i].id == NETSS_SERVICE_ID_BBU)
+      {
+         netss_indicate_bbu_service_level((service_msg->service[i].state == NETSS_SERVICE_POWER_LEVEL_ON)? 1:0);
+      }
       if(!g_netss_boot_handshake_done)
       {
          /*Service completed initalized by moving to the requested power level*/
